backend_new/src/services/listing.service.ts

```diff
@@ -0,0 +1,330 @@
+
+  /*
+    this will get a set of listings given the params passed in
+    this set can either be paginated or not depending on the params
+    it will return both the set of listings, and some meta information to help with pagination
+  */
+  async list(params: ListingsQueryParams) {
+    const whereClause = this.buildWhereClause(params.filter, params.search);
+    const isPaginated = shouldPaginate(params.limit, params.page);
+
+    const count = await this.prisma.listings.count({
+      where: whereClause,
+    });
+
+    const listingsRaw = await this.prisma.listings.findMany({
+      skip: calculateSkip(params.limit, params.page),
+      take: calculateTake(params.limit),
+      orderBy: buildOrderBy(params.orderBy, params.orderDir),
+      include: views[params.view ?? 'full'],
+      where: whereClause,
+    });
+
+    const listings = mapTo(ListingGet, listingsRaw);
+
+    listings.forEach((listing) => {
+      if (Array.isArray(listing.units) && listing.units.length > 0) {
+        listing.unitsSummarized = {
+          byUnitTypeAndRent: summarizeUnitsByTypeAndRent(
+            listing.units,
+            listing,
+          ),
+        };
+      }
+    });
+    const itemsPerPage =
+      isPaginated && params.limit !== 'all' ? params.limit : listings.length;
+    const totalItems = isPaginated ? count : listings.length;
+
+    const paginationInfo = {
+      currentPage: isPaginated ? params.page : 1,
+      itemCount: listings.length,
+      itemsPerPage: itemsPerPage,
+      totalItems: totalItems,
+      totalPages: Math.ceil(
+        totalItems / itemsPerPage ? itemsPerPage : totalItems,
+      ),
+    };
+
+    return {
+      items: listings,
+      meta: paginationInfo,
+    };
+  }
+
+  /*
+    this helps build the where clause for the list()
+  */
+  buildWhereClause(
+    params?: ListingFilterParams[],
+    search?: string,
+  ): Prisma.ListingsWhereInput {
+    const filters: Prisma.ListingsWhereInput[] = [];
+
+    if (params?.length) {
+      params.forEach((filter) => {
+        if (ListingFilterKeys.name in filter) {
+          const builtFilter = buildFilter({
+            $comparison: filter.$comparison,
+            $include_nulls: false,
+            value: filter[ListingFilterKeys.name],
+            key: ListingFilterKeys.name,
+          });
+          filters.push({
+            OR: builtFilter.map((filt) => ({ [ListingFilterKeys.name]: filt })),
+          });
+        } else if (ListingFilterKeys.status in filter) {
+          const builtFilter = buildFilter({
+            $comparison: filter.$comparison,
+            $include_nulls: false,
+            value: filter[ListingFilterKeys.status],
+            key: ListingFilterKeys.status,
+          });
+          filters.push({
+            OR: builtFilter.map((filt) => ({
+              [ListingFilterKeys.status]: filt,
+            })),
+          });
+        } else if (ListingFilterKeys.neighborhood in filter) {
```

Comment:
some repetition in this file w this block but nbd

Reply:
yeah when I was initially working on this I tried to make a generic function to handle all cases and that sorta fell through when I hit the bedrooms filter and looking ahead to detroit filters

I'm open to suggestions though, should I still make a generic function to help with this stuff and handle those edge cases separately?